### PR TITLE
rgw: continuation of the auth rework

### PR DESCRIPTION
**Pull request by rzarzynski**
Marking as *DNM* because:
* S3 errors codes should be differentiated - now we're mostly returning just `AccessDenied` while s3-tests require more detailed info (eg. `SignatureDoesNotMatch`),
* some auth engines (`rgw::auth::s3::LDAPEngine`, `rgw::auth::swift::SignedTokenEngine`, `rgw::auth::swift::ExternalTokenEngine` and so on) need to be tested,
* rebase to `master` is necessary.

CC: @mattbenjamin, @cbodley, @pritha-srivastava.

### CODE REVIEW

**Comment by rzarzynski** [comment]
@pritha-srivastava: I've added https://github.com/ceph/ceph/pull/12893/commits/ea6c1d175eba270f32f318bb7016dc7d6ad50914. Now the s3-tests look much better.

**Comment by mattbenjamin** [comment]
@rzarzynski I think merge is relatively trivial, but we should still rebase?

**Comment by mattbenjamin** [comment]
git@github.com:linuxbox2/ceph.git
rzarzynski-wip-rgw-auth-rework-cont-2

File: src/rgw/rgw_auth.h [hunk 1]
```diff
@@ -101,17 +104,225 @@ class RGWAuthApplier : public RGWIdentityApplier {
...
+      : reason(reason) {
+    }
+
+    AuthResult(rejection_mark_t&&, const int reason)
+      : is_rejected(true),
+        reason(reason) {
+    }
+
+    /* Allow only the reasonable combintations - returning just Completer
+     * without ccompanying IdentityApplier is strictly prohibited! */
```
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: FIXME: typo.

File: src/rgw/rgw_auth.h [hunk 2]
```diff
@@ -101,17 +104,225 @@ class RGWAuthApplier : public RGWIdentityApplier {
...
+    }
+
+    AuthResult(IdentityApplier::aplptr_t&& applier,
+               Completer::cmplptr_t&& completer)
+      : result_pair(std::move(applier), std::move(completer)) {
+    }
+
+  public:
+    enum class Status {
+      /* Engine does */
```
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: FIXME: this **must** be documented.

File: src/rgw/rgw_auth.h [hunk 3]
```diff
@@ -101,17 +104,225 @@ class RGWAuthApplier : public RGWIdentityApplier {
...
+     * will be tried. */
+    REQUISITE,
+
+    /* Success of an engine injected with the SUFFICIENT specifier ends
+     * strategy's authentication process successfully. However, failure
+     * doesn't abort it -- there will be fall-back to following engine
+     * it the one that failed wasn't the last one. */
+    SUFFICIENT,
+
+    /* */
```
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: FIXME: document this.

File: src/rgw/rgw_auth.h [hunk 4]
```diff
@@ -138,32 +349,37 @@ class RGWRemoteAuthApplier : public RGWAuthApplier {
     }
   };
 
-  using aclspec_t = RGWIdentityApplier::aclspec_t;
+  using aclspec_t = rgw::auth::Identity::aclspec_t;
   typedef std::function<uint32_t(const aclspec_t&)> acl_strategy_t;
 
 protected:
+  CephContext * const cct;
```
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: FIXME: move the `*` sign.

File: src/rgw/rgw_auth_filters.h [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+/* Abstract decorator over any implementation of rgw::auth::IdentityApplier
+ * which could be provided both as a pointer-to-object or the object itself. */
+template <typename DecorateeT>
+class DecoratedApplier : public rgw::auth::IdentityApplier {
+  typedef typename std::remove_pointer<DecorateeT>::type DerefedDecorateeT;
+
+  static_assert(std::is_base_of<rgw::auth::IdentityApplier,
+                                DerefedDecorateeT>::value,
+                "DecorateeT must be a subclass of rgw::io::RestfulClient");
```
Comment (rzarzynski) [inline src/rgw/rgw_auth_filters.h]: FIXME: oops, wrong assert message.

File: src/rgw/rgw_auth_filters.h [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+  DecorateeT decoratee;
+
+  /* There is an indirection layer over accessing decoratee to share the same
+   * code base between dynamic and static decorators. The difference is about
+   * what we store internally: pointer to a decorated object versus the whole
+   * object itself. */
+  template <typename T = void,
+            typename std::enable_if<
```
Comment (rzarzynski) [inline src/rgw/rgw_auth_filters.h]: Hmm, `std::enable_if<>` & friends have a a great potential to confuse people. Maybe just adding _SFINAE_ (or similar keyword) to the comment above could make googling easier.

File: src/rgw/rgw_auth_filters.h [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    /* There is no override specified by the upper layer. This means that we'll
+     * load the account owned by the authenticated identity (aka auth_user). */
+    DecoratedApplier<T>::load_acct_info(user_info);
+  } else if (DecoratedApplier<T>::is_owner_of(acct_user_override)) {
+    /* The override has been specified but the account belongs to the authenticated
+     * identity. We may safely forward the call to a next stage. */
+    DecoratedApplier<T>::load_acct_info(user_info);
+  } else {
+    /* Compatibility mechanism for multi-tenancy. For more details refer to
+     * load_acct_info method of RGWRemoteAuthApplier. */
```
Comment (rzarzynski) [inline src/rgw/rgw_auth_filters.h]: FIXME: update `RGWRemoteAuthApplier`.

File: src/rgw/rgw_auth_keystone.cc [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+TokenEngine::result_t
+TokenEngine::authenticate(const std::string& token,
+                          const req_state* const s) const
+{
+  boost::optional<TokenEngine::token_envelope_t> t;
+
+  /* This will be initialized on the first call to this method. In C++11 it's
+   * also thread-safe. */
+  static const struct RolesCacher {
+    RolesCacher(CephContext * const cct) {
```
Comment (rzarzynski) [inline src/rgw/rgw_auth_keystone.cc]: FIXME: move the `*` sign.

File: src/rgw/rgw_auth_keystone.cc [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+  int ret = rgw::keystone::Service::get_admin_token(cct, token_cache, config,
+                                                    admin_token);
+  if (ret < 0) {
+    ldout(cct, 2) << "s3 keystone: cannot get token for keystone access"
+                  << dendl;
+    throw ret;
+  }
+
+  using RGWValidateKeystoneToken
+    = rgw::keystone::Service::RGWValidateKeystoneToken;
```
Comment (rzarzynski) [inline src/rgw/rgw_auth_keystone.cc]: This looks ugly. Maybe something like:
```
  using RGWValidateKeystoneToken = \
    rgw::keystone::Service::RGWValidateKeystoneToken;
```

File: src/rgw/rgw_auth_keystone.cc [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+  /* set required headers for keystone request */
+  validate.append_header("X-Auth-Token", admin_token);
+  validate.append_header("Content-Type", "application/json");
+
+  /* check if we want to verify keystone's ssl certs */
+  validate.set_verify_ssl(cct->_conf->rgw_keystone_verify_ssl);
+
+  /* create json credentials request body */
+  JSONFormatter credentials(false);
+  credentials.open_object_section("");
```
Comment (rzarzynski) [inline src/rgw/rgw_auth_keystone.cc]: TODO: clean-up to a separated function.

File: src/rgw/rgw_json_enc.cc [hunk 11]
```diff
@@ -1325,20 +1300,29 @@ void rgw_sync_error_info::dump(Formatter *f) const {
   encode_json("message", message, f);
 }
 
-void KeystoneAdminTokenRequestVer2::dump(Formatter * const f) const
+/* This utility function shouldn't conflict with the overload of std::to_string
+ * provided by string_ref since Boost 1.54 as it's defined outside of the std
+ * namespace. I hope we'll remove it soon - just after merging the Matt's PR
```
Comment (rzarzynski) [inline src/rgw/rgw_json_enc.cc]: The PR is merged. :-)

File: src/rgw/rgw_auth_keystone.cc (hunk 9 above)
Comment (cbodley) [inline src/rgw/rgw_auth_keystone.cc]: if you're not changing the name itself, just:
```c++
using rgw::keystone::Service::RGWValidateKeystoneToken;
```

Comment (rzarzynski) [inline src/rgw/rgw_auth_keystone.cc]: This is much, much better! I've completely forgotten about this possibility. Thanks, Casey!

File: src/rgw/rgw_op.cc [hunk 12]
```diff
@@ -3853,8 +3853,8 @@ int RGWCopyObj::verify_permission()
     }
 
     /* admin request overrides permission checks */
-    if (!s->auth_identity->is_admin_of(src_policy.get_owner().get_id()) &&
-        !src_policy.verify_permission(*s->auth_identity, s->perm_mask,
+    if (! s->auth.identity->is_admin_of(src_policy.get_owner().get_id()) &&
+        ! src_policy.verify_permission(*s->auth.identity, s->perm_mask,
                                       RGW_PERM_READ)) {
```
Comment (rzarzynski) [inline src/rgw/rgw_op.cc]: FIXME: indentation.

File: src/rgw/rgw_rest_s3.cc [hunk 13]
```diff
@@ -1797,141 +1798,65 @@ int RGWPostObj_ObjStore_S3::get_params()
...
-    string s3_access_key;
-    if (!part_str("AWSAccessKeyId", &s3_access_key)) {
+    if (! part_str("AWSAccessKeyId", &s->auth.s3_postobj_creds.access_key)) {
       ldout(s->cct, 0) << "No S3 access key found!" << dendl;
       err_msg = "Missing access key";
       return -EINVAL;
     }
     string received_signature_str;
-    if (!part_str("signature", &received_signature_str)) {
+    if (!part_str("signature", &s->auth.s3_postobj_creds.signature)) {
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: FIXME:
```
if (! part_str("signature", &s->auth.s3_postobj_creds.signature)) {
```

File: src/rgw/rgw_rest_s3.cc [hunk 14]
```diff
@@ -4463,19 +4258,21 @@ void RGWLDAPAuthEngine::init(CephContext* const cct)
...
   return nullptr;
 }
 
-RGWRemoteAuthApplier::AuthInfo
-RGWLDAPAuthEngine::get_creds_info(const rgw::RGWToken& token) const noexcept
+rgw::auth::RemoteApplier::AuthInfo
+rgw::auth::s3::LDAPEngine::get_creds_info(const rgw::RGWToken& token) const noexcept
 {
-  using acct_privilege_t = RGWRemoteAuthApplier::AuthInfo::acct_privilege_t;
+  using acct_privilege_t = \
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: FIXME:
```c++
using rgw::auth::RemoteApplier::AuthInfo::acct_privilege_t;
```

**Comment by mattbenjamin** [comment]
@rzarzynski are any of the above changes more than cosmetic?  RGWRemoteAuthApplier?  What is the scope of missing behavior?

File: src/rgw/rgw_auth.h [hunk 15]
```diff
@@ -348,4 +350,173 @@ class RGWAnonymousAuthEngine : public RGWTokenBasedAuthEngine {
...
+ * An engine is supposed only to authenticate (not authorize!) requests
+ * basing on their req_state and provide an upper layer with:
+ *  - rgw::auth::IdentityApplier to commit all changes to the request state as
+ *    well as to the RADOS store (creating an account, synchronizing
+ *    user-related information with external databases and so on).
+ *  - rgw::auth::Completer (optionally) to finish the authentication
+ *    of the request. Typical use case is verifying message integrity
+ *    in AWS Auth v4 and browser uploads (RGWPostObj).
+ *
+ * The authentication process consists two steps:
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth.h]: "consists of"

File: src/rgw/rgw_auth.h [hunk 16]
```diff
@@ -348,4 +350,173 @@ class RGWAnonymousAuthEngine : public RGWTokenBasedAuthEngine {
...
+ *  - Engine::authenticate() supposed to be called before *initiating*
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth.h]: s/supposed/should be/

File: src/rgw/rgw_auth.h [hunk 17]
```diff
@@ -348,4 +350,173 @@ class RGWAnonymousAuthEngine : public RGWTokenBasedAuthEngine {
...
+ *    any modifications to RADOS store that are related to an operation
+ *    a client wants to perform (RGWOp::execute).
+ *  - Completer::complete() supposed to be called, if completer has been
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth.h]: ditto

File: src/rgw/rgw_auth.h [hunk 18]
```diff
@@ -348,4 +350,173 @@ class RGWAnonymousAuthEngine : public RGWTokenBasedAuthEngine {
...
+  enum class Control {
+    /* Failure of an engine injected with the REQUISITE specifier aborts
+     * the whole authentication process immediately. No other engine will
+     * be tried. */
+    REQUISITE,
+
+    /* Success of an engine injected with the SUFFICIENT specifier ends
+     * the whole authentication process successfully. However, failure
+     * doesn't abort it - there will be fall-back to following engine
+     * it the one that failed wasn't the last. */
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth.h]: s/it/if/

File: src/rgw/rgw_auth.cc [hunk 19]
```diff
@@ -509,3 +509,50 @@ RGWAuthApplier::aplptr_t RGWKeystoneAuthEngine::authenticate() const
...
+    const auto& policy = kv.second;
+
+    rgw::auth::Engine::result_t res;
+    try {
+      res = engine.authenticate(s);
+    } catch (int err) {
+      /* NOP */
+    }
+
+    const auto& applier = res.first;
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth.cc]: a bit inexplicit that res.first can be relied on if engine.authenticate threw an exception?

File: src/rgw/rgw_auth.cc [hunk 20]
```diff
@@ -509,3 +509,50 @@ RGWAuthApplier::aplptr_t RGWKeystoneAuthEngine::authenticate() const
...
+auth_fail:
+  /* Returning nullptr as the rgw::auth::Applier means access denied. */
+  return Engine::result_t(nullptr, nullptr);
+}
+
+void
+rgw::auth::Strategy::add_engine(const Control ctrl_flag,
+                                const Engine& engine) noexcept
+{
+  auth_stack.push_back(std::make_pair(std::cref(engine), ctrl_flag));
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth.cc]: interesting, first time I've seen std::cref used

**Comment by mattbenjamin** [comment]
@rzarzynski there's a number of rgw suite failures in this run;  there seem to be only a few root causes, but I haven't dug beyond the teuthology.logs as yet;  I will attempt to close one or more tomorrow--if you can beat me to any, that will be welcome :)
http://pulpito.ceph.com/mbenjamin-2017-03-16_19:38:09-rgw-wip-rzarzynski-rework-c2---basic-smithi/

**Comment by rzarzynski** [comment]
@mattbenjamin: I'm working on that.

**Comment by rzarzynski** [comment]
FIXME: `RGWSignedTokenAuthEngine` -> `rgw::auth::swift:SignedTokenEngine` in debugs.

**Comment by mattbenjamin** [comment]
@rzarzynski help parsing that?

File: src/rgw/rgw_swift_auth.h [hunk 21]
```diff
@@ -7,122 +7,244 @@
...
+                    const rgw::auth::TokenExtractor* const extractor,
+                    const rgw::auth::LocalApplier::Factory* const apl_factory)
+    : cct(cct),
       store(store),
+      extractor(extractor),
       apl_factory(apl_factory) {
   }
 
   const char* get_name() const noexcept override {
     return "RGWSignedTokenAuthEngine";
```
Comment (rzarzynski) [inline src/rgw/rgw_swift_auth.h]: @mattbenjamin: I meant this thing. `get_name()` is used in debugs only. I've forgotten to alter it to reflect the namespaces we have now. 

I will refactor that and push together with rest of similar things.

Comment (mattbenjamin) [inline src/rgw/rgw_swift_auth.h]: ok

**Comment by rzarzynski** [comment]
The failure in Swift functional tests seems to be fixed. Working on s3-tests.

**Comment by rzarzynski** [comment]
It looks that https://github.com/ceph/ceph/commit/ce2b998d3254f0a91be67152984144a9ab1b6f8d resolves the issue with s3-tests according to local testing results.

I've just pushed fixed version of the branch for gitbuilders and will schedule a Teuthology run soon.

**Comment by mattbenjamin** [comment]
@rzarzynski can you get a rebase in? :)

**Comment by rzarzynski** [comment]
@mattbenjamin: I'm working on the last (I hope :-)) bug which popped out during [the yesterday's Teuthology run](http://pulpito.ceph.com/rzarzynski-2017-03-18_19:21:26-rgw-wip-rgw-auth-rework-cont-2-fixups---basic-mira/). It looks that both Swift and S3 auth error codes are handled properly now but we still need to take care of the dynamic reconfiguration in the multi-site scenario. Basically it's all about updating the `RGWRados*` in strategies.

I think I will have the patch till tomorrow's morning.

**Comment by rzarzynski** [comment]
@mattbenjamin: I've just pushed [the patch bringing dynamic reconfiguration of auth strategies](https://github.com/ceph/ceph/commit/e3c3618434ce030d4f3227d5ecc9989133090efd) for gitbuilders. If Teuthology don't find further issues, I will clean-up the fix (`rgw::auth::registry::StrategyRegistry` is definitely too long) and other things as well (the cosmetics + documentation).

**Comment by rzarzynski** [comment]
Unfortunately, gitbuilders [are experiencing](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=trusty,DIST=trusty,MACHINE_SIZE=huge/1888//consoleFull) some issues now. All builds fail because of:
```
Traceback (most recent call last):
  File "/tmp/venv.IkgllbGVns/bin/chacractl", line 6, in <module>
    main.ChacraCtl()
  File "/tmp/venv.IkgllbGVns/local/lib/python2.7/site-packages/chacractl/main.py", line 38, in __init__
    self.main(argv)
  File "/tmp/venv.IkgllbGVns/local/lib/python2.7/site-packages/chacractl/decorators.py", line 68, in newfunc
    return f(*a, **kw)
  File "/tmp/venv.IkgllbGVns/local/lib/python2.7/site-packages/chacractl/main.py", line 68, in main
    self.api_credentials()
  File "/tmp/venv.IkgllbGVns/local/lib/python2.7/site-packages/chacractl/main.py", line 51, in api_credentials
    conf_module = util.load_config()
  File "/tmp/venv.IkgllbGVns/local/lib/python2.7/site-packages/chacractl/util.py", line 23, in load_config
    return imp.load_source('chacractl', config)
  File "/home/jenkins-build/.chacractl", line 1
    url = "<html>
                ^
SyntaxError: EOL while scanning string literal
```
Waiting for them to become operational.

**Comment by rzarzynski** [comment]
@mattbenjamin: the Teuthology runs look good with the exception to non-related issues:
* http://pulpito.ceph.com/rzarzynski-2017-03-20_23:50:08-rgw-wip-rgw-auth-rework-cont-2-fixups---basic-mira/ (two tests failed because of a problem with resolver/`getaddrinfo`),
* http://pulpito.ceph.com/rzarzynski-2017-03-21_00:02:02-rgw-wip-rgw-auth-rework-cont-2-fixups---basic-mira/ (one red test due to Valgrind/`md_config_t`).

I'm addressing the cosmetics and will push final branch soon!

**Comment by mattbenjamin** [comment]
@rzarzynski sounds good and looks good;  with a rebase, this should be ready

File: src/rgw/rgw_auth.h (hunk 1 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_auth.h (hunk 2 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_auth.h (hunk 3 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_auth.h (hunk 4 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_auth_filters.h (hunk 5 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth_filters.h]: DONE.

File: src/rgw/rgw_auth_filters.h (hunk 6 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth_filters.h]: DONE.

File: src/rgw/rgw_auth_filters.h (hunk 7 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth_filters.h]: DONE.

File: src/rgw/rgw_auth_keystone.cc (hunk 8 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth_keystone.cc]: DONE.

File: src/rgw/rgw_auth_keystone.cc (hunk 9 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth_keystone.cc]: DONE.

File: src/rgw/rgw_op.cc (hunk 12 above)
Comment (rzarzynski) [inline src/rgw/rgw_op.cc]: DONE.

File: src/rgw/rgw_rest_s3.cc (hunk 13 above)
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: DONE.

File: src/rgw/rgw_rest_s3.cc (hunk 14 above)
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: DONE: commented that the short form of `using` can't be used here.

File: src/rgw/rgw_auth.h (hunk 15 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_auth.h (hunk 16 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_auth.h (hunk 18 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.h]: DONE.

File: src/rgw/rgw_swift_auth.h (hunk 21 above)
Comment (rzarzynski) [inline src/rgw/rgw_swift_auth.h]: DONE.

File: src/rgw/rgw_auth.cc (hunk 19 above)
Comment (rzarzynski) [inline src/rgw/rgw_auth.cc]: This has been superseded later by:
```cpp
    result_t engine_result = result_t::deny();
    try {
      engine_result = engine.authenticate(s);
    } catch (const int err) {
      engine_result = result_t::deny(err);
    }
```

File: src/rgw/rgw_auth.cc [hunk 22]
```diff
@@ -210,38 +364,38 @@ void RGWRemoteAuthApplier::load_acct_info(RGWUserInfo& user_info) const      /*
...
       << ", acct_name=" << user_info.display_name
       << ", subuser=" << subuser
       << ", perm_mask=" << get_perm_mask()
-      << ", is_admin=" << user_info.admin << ")";
+      << ", is_admin=" << static_cast<bool>(user_info.admin) << ")";
 }
 
-uint32_t RGWLocalAuthApplier::get_perm_mask(const std::string& subuser_name,
-                                            const RGWUserInfo &uinfo) const
+uint32_t rgw::auth::LocalApplier::get_perm_mask(const std::string& subuser_name,
```
Comment (mikulely) [inline src/rgw/rgw_auth.cc]: hi @rzarzynski  can we  parse  `perm_mask` and `subuser_name` once? Caz we've a requirement that need to find subuser_name during auth phrase to impl [subuser usage log statistics](https://github.com/ceph/ceph/pull/13929) .

Comment (rzarzynski) [inline src/rgw/rgw_auth.cc]: @mikulely: I posted a comment regarding that in the mentioned pull request.

**Comment by rzarzynski** [comment]
@mattbenjamin: just pushed the rebased branch with fixes squashed in. I will schedule a Teuthology run soon.

**Comment by rzarzynski** [comment]
Scheduled a Teuthology run:
* http://pulpito.ceph.com/rzarzynski-2017-03-25_11:22:50-rgw-wip-rgw-auth-rework-cont-2---basic-smithi/.

**Comment by rzarzynski** [comment]
@mattbenjamin: because of the very long queue on `smithi`, I've scheduled [the additional run on `mira`](http://pulpito.ceph.com/rzarzynski-2017-03-25_18:37:43-rgw-wip-rgw-auth-rework-cont-2---basic-mira/). Generally speaking the results look good. Following unrelated failures were found:
* one job failed because `[WRN] message from mon.0 was stamped 0.600861s in the future, clocks not synchronized`,
* one job thrown in `s3-tests` because of the well-known `100 Continue` issue on Apache and FastCGI front-end,
* in one job RadosGW instance finished due to a multi-site issue: `2017-03-25 22:00:58.828106 309d4700  0 ERROR: failed to clone shard, completion_mgr.get_next() returned ret=-125`. Errno `125` is `ECANCELED`,
* one job is red due because `No summary info found for user: foo`. I think I saw this message couple of times on very different branches,
* two jobs failed as Valgrind detected the well-known problems with `md_config_t`.

**Comment by rzarzynski** [comment]
@mattbenjamin: [the run on `smithi`](http://pulpito.ceph.com/rzarzynski-2017-03-25_11:22:50-rgw-wip-rgw-auth-rework-cont-2---basic-smithi/) finished very recently and it's even greener. Only following non-related issues:
* two jobs failed on Apache with FastCGI in `testChunkedPut (test.functional.tests.TestFileUTF8)` of the Swift's tests. [The same affects `master`](http://pulpito.ceph.com/teuthology-2017-03-25_03:05:02-rgw-master---basic-smithi/942783/) so we don't have a regression here,
* one job failed as Valgrind detected the well-known problem with `md_config_t`.

**Comment by mattbenjamin** [comment]
@rzarzynski ack, well done

**Comment by cbodley** [comment]
great work on this @rzarzynski! thanks for being patient with us in our design critique, i think it turned out really well

**Comment by rzarzynski** [comment]
@mattbenjamin, @cbodley: Guys, it's really nice hear but let's wait a bit. We still have AWSv4 to address. :-)

**Pull request merged**